add `@mock` spec (#10)
graphql/gaps: GAP-0/DRAFT.md

  business(id: "123") {
    name
    hours @mock(variant: "morning-only") {
      open
      close
    }
  }
}

# mock the entire operation
query GetBusinessRating @mock(variant: "five-star-bakery") {
  business(id: "123") {
    name
    rating
  }
}
```

Mock data is stored in JSON files and read statically when requests are
executed:

```json example
{
  "morning-only": {
    "data": {
      "open": "8:00am",
      "close": "12:00pm"
    },
    "__path__": "business.hours"
  }
}
```

The client transforms the document to remove mocked
_[selections](https://spec.graphql.org/September2025/#Selection)_ before
executing or sending the request to the server. Upon receiving a response from
the server, mock values are merged into the response object before yielding to
the application.

Use of Large Language Models (LLMs) is suggested as a means of creating and
maintaining mock data.

# Directive

```graphql
directive @mock(
  variant: String
  value: String
) on QUERY | MUTATION | SUBSCRIPTION | FIELD
```

## Arguments

`@mock` accepts two mutually exclusive arguments: {"variant"} and {"value"}.
Exactly one of these arguments must be provided. Providing both {"variant"} and

> {"value"} in the same `@mock` application is a validation error.

Reviewer: ```suggestion
{"value"} in the same `@mock` directive is a validation error.
```

*mock value* (stored in the *mock file*) to return for the field or operation
where `@mock` is applied.

### value

{"value"} provides an inline *mock value* directly in the document.

When {"value"} is provided, the client uses the supplied string as the mock
response for the annotated field without consulting a *mock file*.

{"value"} may only be applied to fields that resolve to
_[leaf types](https://spec.graphql.org/September2025/#sec-Leaf-Field-Selections)_
(scalars and enums). It must not be applied to fields that return object types
or to operation roots.

Because {"value"} is always a string, the client must coerce the string to the
appropriate JSON type before merging it into the response:

CoerceInlineValue(value) :
  1. If {value} is {"null"}, return `null`.
  1. If {value} is {"true"}, return the boolean `true`.
  1. If {value} is {"false"}, return the boolean `false`.
  1. If {value} can be parsed as a base-10 number, return {value} as number.
  1. Return {value} as a string.

Note: This specification requires JSON as the serialization format for GraphQL
responses.

## Returning Mock Data

If `@mock` is applied to an operation definition (e.g. {"Query"}), the entire

> response must be resolved from a *mock file*.

Reviewer: ```suggestion
response must be resolved from a *mock file*; no request should be sent to the server.
```

If `@mock` is applied to non-root fields only, the client must transform the
document to remove any selections which have `@mock` applied before sending the
request to the server. For fields using {"variant"}, mock values are resolved
from a *mock file*. For fields using {"value"}, the mock value is the coerced
inline string. When the server's response is received, the client merges each
*mock value* into the response before yielding to the application.

When `@mock` is applied to a field, the *mock value* must always be included in
the response, regardless of other directives present on the same field. In
particular, `@mock` takes precedence over `@skip` and `@include` — the mock
data is always returned, irrespective of whether the field would otherwise be
skipped or included based on those directives' conditions.

Mock values must **not** be generated dynamically at runtime. Mock values must
be resolved from the *mock file* or from the inline {"value"} argument.

The mechanism for GraphQL clients running in a web browser or mobile client to

> read the *mock file* is implementation defined. See:

Reviewer: ```suggestion
read the *mock file* is implementation-defined. See:
```

[Appendix: Reading Mock Files](#sec-Reading-Mock-Files).

# Mock Files

:: A *mock file* is a `.json` file that maps each *mock variant id* to a
*mock variant*.

Each operation or fragment that contains one or more `@mock` directives using
the {"variant"} argument must have an associated *mock file*. Operations or
fragments where all `@mock` directives use only the {"value"} argument do not
require a *mock file*.

Note: Mock files are intended to be long-lived and may be checked into version
control. This is useful for client developers working on a project over an
extended period of time, and where the client code depends on GraphQL schema
that does not yet exist.

## Mock File Location

The *mock file* must be named `{Name}.json`, where `{Name}` is the name of the

> operation or fragment for which the mock values may be used.

Reviewer: ```suggestion
operation or fragment within which the mock values may be used.
```

To avoid mock file naming collisions, this specification requires that all
operation and fragment names within a project are unique.

The mock file must be stored in a directory named {"__graphql_mocks__"} adjacent
to the source file containing the operation or fragment.

**Example**

For an operation named `GetBusinessInfo` defined in `BusinessDetails.graphql`:

```
.
├── __graphql_mocks__
│   └── GetBusinessInfo.json
└── BusinessDetails.graphql
```

TODO: Support multiple files per operation/fragment. Instead of {Name}.json, it
may be a directory of mock files, nested by the operation/fragment name. e.g.

> `{Name}/arbitrary.json`.

Reviewer: Can you elaborate on what this would solve? Curious as to what the gap in the current proposal is without this

## Mock File Structure

The mock file contains a JSON object which maps *mock variant id* keys to a
*mock variant*.

:: A *mock variant id* is any key in the object that does not start with two
underscores (`__`).

:: A *mock variant* is the object associated with each *mock variant id*.

### Mock Variant Object

A *mock variant* object may contain **only** the following keys:

- {"data"} (required)
- {"errors"}
- {"extensions"}
- {"__path__"} (required)
- {"__description__"}
- {"__metadata__"}

#### data

> :: {"data"} stores the *mock value*. It may be `null` if the field is nullable.

Reviewer: Might be worth calling out that this can be an object or scalar value, depending on what is being mocked

#### errors

{"errors"} may contain an array of error objects conforming to the
[GraphQL error format](https://spec.graphql.org/September2025/#sec-Errors).

> This must be merged into the GraphQL operation's response if defined.

Reviewer: I think it'd be nice to clarify whether this is valid _only_ for `@mock` on operations or both for operations _and fields_

> TODO: Define merging algorithm and resolution algorithm for conflicts

Reviewer: Might not need this since each field can have multiple errors in the spec

#### extensions

{"extensions"} may be a key-value mapping of arbitrary data. This must be merged

> into the GraphQL operation's response if defined.

Reviewer: Probably want to call out that they need to be compliant with https://spec.graphql.org/September2025/#sec-Extensions

TODO: Define merging algorithm and resolution algorithm for conflicts

#### __description__

{"__description__"} may contain a string which describes the *mock value* in
natural language. This value should be used when regenerating the *mock value*.

```json example
{
  "5-star-business": {
    "data": {
      "business": {
        "name": "The Great British Bakery",
        "rating": 5.0
      }
    },
    "__path__": "Query",
    "__description__": "A delicious bakery with a rating of 5.0"
  }
}
```

#### __path__

{"__path__"} is the *field path* within the operation or fragment where `@mock`
is or may be applied for a given *mock variant id*.

:: A *field path* is a dot-separated string of field names (or aliases, where
present) representing the location of the field relative to the root of the
operation or fragment. 

TODO: Define field paths in a separate specification. See
https://github.com/graphql/graphql-spec/issues/1215

For `@mock` on an operation root, {"__path__"} is the root operation type name

> (e.g. {"Query"}).

Reviewer: Is this also true for `Mutation` and `Subscription`?

**Example**

For the following fragment:

```graphql example
fragment FooFields on Foo {
  # field path is "bar"
  bar @mock(variant: "basic-bar")

  # field path is "aliasedBar"
  aliasedBar: bar @mock(variant: "aliased-bar")

  baz {
    # field path is "baz.qux"
    qux @mock(variant: "basic-qux")
  }
}
```

This would be a (minimally) valid corresponding *mock file*:

```json example
{
  "basic-bar": {
    "data": "...",
    "__path__": "bar"
  },
  "aliased-bar": {
    "data": "...",
    "__path__": "aliasedBar"
  },
  "basic-qux": {
    "data": "...",
    "__path__": "baz.qux"
  }
}
```

#### __metadata__

{"__metadata__"} may be a key-value mapping for additional user or application
defined metadata.

# Validation

As development progresses, the shape of the operation or schema may change.
Since mock files may be checked into version control and persist across schema
changes, an existing *mock value* in a *mock file* or inline {"value"} argument
may become invalid over time.

> Conforming clients must check that mock data is valid for each operation.

Reviewer: ```suggestion
Conforming clients must verify that mock data is valid for each operation at runtime.
```

## Mock File Validation

If a *mock variant id* referenced by a {"variant"} argument does not exist in
the *mock file*, this is a validation error.

> All *mock variant ids* within a *mock file* must be unique; duplicate keys are a

Reviewer: ```suggestion
All *mock variant ids* within a *mock file* must be unique; duplicate keys trigger a
```

validation error.

GraphQL clients must raise an error for an invalid *mock value* defined inside
a *mock file*.

A *mock value* is valid when its shape is compatible with the operation's
selections at the *field path* where `@mock` is applied. For each selected
field, the *mock value* must satisfy
_[CompleteValue](https://spec.graphql.org/draft/#CompleteValue())_ for the
field's schema type. Fields present in the operation but not defined in the
schema are skipped during validation.

Note: It is also possible to detect if a JSON payload is valid for a given
operation by constructing an in-memory GraphQL server that has no resolvers, and
uses the JSON payload as its {rootValue} — and ensuring no errors are thrown for
execution of the operation against the test server. The schema must be modified

> to include any new types and fields referenced in the *mock value*.

Reviewer: ```suggestion
Note: It is also possible to detect if a JSON payload is valid for a given
operation by constructing an in-memory GraphQL server that has no resolvers and
uses the JSON payload as its {rootValue} — then ensuring no errors are thrown for
execution of the operation against the test server. The schema must be modified
to include any new types and fields referenced in the *mock value*.
```

## Inline Mock Value Validation

The {"value"} argument may only be used on scalar fields. When the field exists

> in the schema, the coerced value must match the field's return type.

Reviewer: I suppose we could also allow this for enum values, thoughts?

# Appendix

_This section is non-normative._

## Dynamic Mock Variant IDs

The {"variant"} argument could be passed in from operation
_[variables](https://spec.graphql.org/September2025/#sec-Language.Variables)_.

A use case would be to build an interface to let viewers control which
*mock value* is used. For example, the following GET parameters:

```example
GET /my_page?gql_mock_variants=barMock=some-variant-id;bazMock=some-other-variant-id
```

This could be mapped to operation variables for the following operation:

```graphql example
query Foo($barMock: String, $bazMock: String) {
  bar @mock(variant: $barMock)
  baz @mock(variant: $bazMock)
}
```

Clients must manually coerce the directive argument variables (similar to
[CoerceArgumentValues](https://spec.graphql.org/draft/#CoerceArgumentValues())).

This is currently outside the scope of this specification.

## Reading Mock Files

GraphQL clients that implement this specification must have access to the data
contained in mock files when requests are executed. Since web browsers and
mobile applications cannot read remote file systems, a mechanism is required
to make the data contained in mock files available to the client.

> This mechanism is implementation defined.

Reviewer: ```suggestion
This mechanism is implementation-defined.
```

## Schema-Aware Clients

Schema-aware clients face additional complexity: operations using the `@mock`
directive may reference types and fields not yet present in the server schema.
Such clients must patch their local schema to include these definitions, or
disable validation for mocked operations. The mechanism for schema patching is
outside the scope of this specification, however contributions addressing this
are welcome for inclusion in a future version of this document.

**Trusted Documents**

TODO: add a section to clarify that while this intended to be used in local dev
(where the documents don't need to be hashed), because mock files can get
checked in, and @mock directives could get checked in, @mock selections need
to not be included in the actual document that gets hashed (i.e. the same
transforming needs to happen at runtime and query compile time)

> This could be useful in staging environments or on prod (behind a feature flag). 

Reviewer: ```suggestion
TODO: Add a section to clarify that while this is intended to be used in local development
(where the documents don't need to be hashed), because mock files can get
checked in, and @mock directives can get checked in, @mock selections need
to be excluded from the actual document that gets hashed (i.e. the same
transforming needs to happen at runtime and query compile time)
This could be useful in staging environments or in production (behind a feature flag). 
```

{"value"} may only be applied to fields that resolve to
_[leaf types](https://spec.graphql.org/September2025/#sec-Leaf-Field-Selections)_
(scalars and enums). It must not be applied to fields that return object types
or to operation roots. The client must coerce the string value to the field's
expected scalar type (for example, `"42"` becomes the integer {42} for an
{Int} field, and `"true"` becomes the boolean {true} for a {Boolean} field).

```graphql example
query GetBusinessInfo {
  business(id: "123") {
    name @mock(value: "The Great British Bakery")
    rating @mock(value: "4.5")
    isOpen @mock(value: "true")
    hours @mock(variant: "morning-only") {
      open
      close
    }
  }
}
```

Note: {"value"} is useful for quick, self-contained mocks of scalar fields
where the overhead of creating and maintaining a *mock file* entry is not
warranted.

## Returning Mock Data

If `@mock` is applied to the operation's root field (e.g. {"query"}), the entire
response must be resolved from a *mock file*. The {"value"} argument must not be
used on operation roots.

If `@mock` is applied to non-root fields only, the client must transform the

> document to remove any selections which have `@mock` applied before sending the

Reviewer: One thing that is probably worth noting here is that for a document to remain valid, the field removal likely needs to happen recursively:
- Empty selection sets are invalid, so if all fields within a selection set are marked with `@mock`, the parent field needs to be removed as well. Same goes for fragment definitions and inline fragments
- Unused variables are invalid, so if a `@mock`ed field is the only field that used an argument, the argument needs to be removed from the root query
- An operation with no selection sets is invalid, `@mock`ing all top-level fields should be disallowed (in favor of mocking the top-level operation itself)

Author: great points.

Reviewer: Looks like these clarifications are still missing; maybe we add them to the next line?

applied.

Mock files are intended to be long-lived and may checked into version control.

Note: This is useful for client developers working on a project over an extended
period of time, and where the client code depends on GraphQL schema that does
not yet exist.

### Mock File Location

The *mock file* must be named `{Name}.json`, where `{Name}` is the name of the
*executable target* (that is, the name of the operation or fragment in which
the mock variants may be applied).

TODO: Support multiple files per target - instead of {Name}.json, it could
be a directory of mock files, nested by the operation target - i.e.
{Name}/arbitrary.json

Mock files should be stored in a `__graphql_mocks__` directory adjacent to the
source file containing the *operation target*.

**Example**

For an operation named `GetBusinessInfo` defined in `BusinessDetails.graphql`:

```
.
├── __graphql_mocks__
│   └── GetBusinessInfo.json
└── BusinessDetails.graphql
```

### Mock File Structure

The mock file contains a JSON object which maps *mock variant id* keys to a
*mock variant*.

:: A *mock variant id* is any key in the object that does not start with two
underscores (`__`).

> :: A *mock vatiant* is the object associated with each *mock variant id*.

Reviewer: Shall we require that all mock variants within a given type/file have unique names to prevent collisions? (This should likely be part of mock validation)

Author: implicitly so but worth calling out!

The *mock variant* object may contain only the following keys:

- {"data"} (required)
- {"errors"}
- {"extensions"}
- {"__appliesTo__"} (required)
- {"__description__"}
- {"__metadata__"}

#### data

> :: {"data"} stores the *mock value*. It may be `null` if the field is nullable.

Reviewer: It might be worth noting that this data will be automatically inserted within the root `data` key in the operation's response data. In other words, for a `@mock` on a root query, its entry would be this:

```json
{
  "mock-id": {
    "data": {
      "business": {
        "name": "The Great British Bakery",
        "rating": 5.0
      }
    },
    ...
  },
}
```

**not** this:

```json
{
  "mock-id": {
    "data": {
      "data" {
        "business": {
          "name": "The Great British Bakery",
          "rating": 5.0
        }
      }
    },
    ...
  },
}
```

#### errors

{"errors"} may contain an array of error objects conforming to the
[GraphQL error format](https://spec.graphql.org/September2025/#sec-Errors).
This must be merged into the GraphQL operation's response if defined.

TODO: Define merging algorithm and resolution algorithm for conflicts

#### extensions

{"extensions"} may be a key-value mapping of arbitrary data. This must be merged
into the GraphQL operation's response if defined.

TODO: Define merging algorithm and resolution algorithm for conflicts

#### ___description__

{"__description__"} may contain a string which describes the *mock value* in
natural language. This value should be used when regenerating the *mock value*.

```json example
{

>   "5-star-business: {

Reviewer: ```suggestion
  "5-star-business": {
```

    "data": {
      "business": {
        "name": "The Great British Bakery",
        "rating": 5.0
      }
    },
    "__appliesTo__": "Query",
    "__description__": "A delicious bakery with a rating of 5.0"
  },
}
```

#### __appliesTo__

{"__appliesTo__"} must be defined as the resolved
_[schema coordinate](https://spec.graphql.org/September2025/#sec-Schema-Coordinates)_
of the field or root operation type where `@mock` is applied, or may be applied,
with the *mock variant id*.

**Example**

For the following fragment:

```graphql example
fragment FooFields on Foo {
  # location resolves to "Foo.bar"
  bar @mock(variant: "basic-bar")

  # location resolves to "Foo.baz"
  baz @mock(variant: "basic-baz") {
    qux
  }
}

> ```

Reviewer: From our call today:
- Using schema coordinates here will not work for field aliases (since those don't exist in the schema)
- We need to know the path for which a given mock is being applied to (not only the type) in order to ensure it provides necessary data for all the required fields (see the example below - `user-1` and `user-2` variants represent the same type, but both would be invalid for the other's field path)
- **To work around this, we propose changing `__appliesTo__` to `__path__` and changing its representation to be a _path relative to the parent executable fragment/operation_**
- For `@mock` on the root query, `__path__` could be `.` `<root>`, or something else

```graphql
query SomeQuery {
  user(id: 123) @mock(variant: 'user-1') { # returns User
    name
  }
  house(id: 123) { # returns User
    owner @mock(variant: 'renter-1') {
      name
      avatarUrl
    }
  }
  somethingElseThatReturnsAUser @mock(variant: 'user-2') { # returns User
    password
  }
}
```

Author: **Option 1**

- `@mock` on a query root does `__path__ = "Query"` 
- `@mock` within queries do `__path__ = "Query.bar.baz"` 
- `@mock` within fragments do `__path__ = "Foo.bar.baz"` (where "Foo" is the name of the fragment)

this feels nice but also possibly redundant, given that the "Query." or "<FragmentName>." prefix is already implied by the mock file name.

**Option 2**

Same as above but remove the "Query." prefix for mocks within an operation/fragment:

- `@mock` on a query root does `__path__ = "Query"` 
- `@mock` within queries do `__path__ = "bar.baz"` 
- `@mock` within fragments do `__path__ = "bar.baz"`

Given that [response positions](https://github.com/graphql/graphql-spec/blob/3c6695ce9cba27a2b887efd1f0e06cb030731b7c/spec/Section%207%20--%20Response.md#response-position) also do not use a "Query." prefix I guess I'll go with option 2, but happy to bikeshed on this.

Author: @benjie points out that https://github.com/graphql/graphql-wg/blob/main/rfcs/OperationExpressions.md is already a (proposed) thing we could use here instead.

Reviewer: Option 2 seems reasonable to me 👍🏽

This would be a (minimally) vald corresponding *mock file*:

```json example
{
  "basic-bar": {
    "data": "...",
    "__appliesTo": "Foo.bar"
  },
  "basic-baz": {
    "data": { "qux": "..." },
    "__appliesTo": "Foo.baz"
  },
}
```

#### __metadata__

{"__metadata__"} may be a key-value mapping for additional user or application
defined metadata.

> ## Mock Entry Validation

Reviewer: I think we also want to disallow nested `@mock` directives since they'll collide. Thoughts?

Author: Makes sense.

Potential frustration:

```graphql
query MockedFoo {
  foo @mock(variant: "foo-fields") {
    ...FooFields
  }
}
```

```graphql
query UnmockedFoo {
  foo {
    ...FooFields
  }
}
```

```graphql
fragment FooFields on Foo {
  bar @mock(value: "baz")
}
```

`bar @mock(value: "baz")` is only invalid in the context of being spread into the `MockedFoo` query, but would be valid when used in `UnmockedFoo`.

So the issues are:

1) slight added complexity: the validation logic must resolve the fragment imports for for each operation in order to fully check for nested `@mock` usage
2) any fragments that get reused in the way shown above cannot use @mock in operations where the parent node isn't mocked. this would be super frustrating.

(2) is the main issue.

But if we had to choose (and I think we do), it's better to frustrate users in an error state that explicitly "doesn't work" and it's obvious why, rather than runtime subtly doing something different to what the user expects (a parent's mock value is applied rather than the child, or vice versa)
